File: backend/app/schemas/frps.py
from pydantic import BaseModel, Field


class FrpsImportRequest(BaseModel):
    dashboard_url: str = Field(min_length=1)
    username: str = Field(min_length=1)
    password: str = Field(min_length=1)
    ssh_port_start: int = 12001
    ssh_port_end: int = 17000
    vnc_port_start: int = 17001
    vnc_port_end: int = 22000
    project_id: int | None = Field(default=None, ge=1)
    location: str | None = "frps"
    overwrite_project_location: bool = False


class FrpsDiscoveredDevice(BaseModel):
    name: str
    device_sn: str
    project_id: int | None
    ssh_port: int
    vnc_port: int | None
    ssh_proxy_name: str
    vnc_proxy_name: str | None
    status: str
    import_status: str = "pending"
    detail: str | None = None
    existing_device_id: int | None = None


class FrpsImportResponse(BaseModel):
    total: int
    created: int
    skipped: int
    synced: int = 0
    conflicts: int = 0
    items: list[FrpsDiscoveredDevice]

File: frontend/src/api/domains/frps.ts
import { api } from "../core";

export interface FrpsImportRequest {
  dashboard_url: string;
  username: string;
  password: string;
  ssh_port_start: number;
  ssh_port_end: number;
  vnc_port_start: number;
  vnc_port_end: number;
  project_id: string;
  location?: string;
  overwrite_project_location?: boolean;
}

export interface FrpsDiscoveredDevice {
  name: string;
  device_sn: string;
  project_id: string;
  ssh_port: number;
  vnc_port: number | null;
  ssh_proxy_name: string;
  vnc_proxy_name: string | null;
  status: string;
  import_status: string;
  detail: string | null;
  existing_device_id: number | null;
}

export interface FrpsImportResponse {
  total: number;
  created: number;
  skipped: number;
  synced: number;
  conflicts: number;
  items: FrpsDiscoveredDevice[];
}

export async function discoverFrpsDevices(payload: FrpsImportRequest): Promise<FrpsImportResponse> {
  const response = await api.post<FrpsImportResponse>("/frps/discover", payload);
  return response.data;
}

export async function importFrpsDevices(payload: FrpsImportRequest): Promise<FrpsImportResponse> {
  const response = await api.post<FrpsImportResponse>("/frps/import", payload);
  return response.data;
}
